### PR TITLE
Add Custom Network UI

### PR DESCRIPTION
**Pull request by brad-decker**
### How To Test Locally
1. Open https://github.com/MetaMask/metamask-extension/blob/eb90cab73c74f91cdedf8d86accd6371e562078a/app/scripts/lib/rpc-method-middleware/handlers/index.js
2. uncomment the commented lines
3. start the extension
4. Make sure you don't have a custom network with `chainId` `0x4d`
5. use the snippet below while on the test-dapp

```js
await ethereum.request({
  method: 'wallet_addEthereumChain',
  params: [{
    chainId: '0x4d',
    rpcUrls: ['https://sokol.poa.network'],
    chainName: 'XDAI (Sokol)',
    nativeCurrency: { symbol: 'XDAI', decimals: 18 },
    blockExplorerUrls: ['https://blockscout.com/poa/sokol']
  }]
})
```

### CODE REVIEW

File: ui/app/pages/confirmation/confirmation.js [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+            .map((alert, idx, filtered) => (
+              <Callout
+                key={alert.id}
+                severity={alert.severity}
+                dismiss={() => dismissAlert(alert.id)}
+                isFirst={idx === 0}
+                isLast={idx === filtered.length - 1}
+                isMultiple={filtered.length > 1}
+              >
+                {React.isValidElement(alert.content) ? (
```
Comment (kumavis) [inline ui/app/pages/confirmation/confirmation.js:228]: on a call we suggested performing this check `React.isValidElement` or a more direct check (`typeof string`) inside `MetaMaskTemplateRenderer`

File: ui/app/pages/confirmation/confirmation.js [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+import { useOriginMetadata } from '../../hooks/useOriginMetadata';
+import { getUnapprovedConfirmations } from '../../selectors';
+import NetworkDisplay from '../../components/app/network-display/network-display';
+import { COLORS, SIZES } from '../../helpers/constants/design-system';
+import Callout from '../../components/ui/callout';
+import { getTemplateValues, getTemplateAlerts } from './templates';
+
+/**
+ * a very simple reducer using produce from Immer to keep state manipulation
+ * immutable and painless. This state is not stored in redux state because it
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/confirmation.js:34]: This seems like a curious decision :thinking: We already have state in Redux that is applicable only to one session and page. I don't think that's a reason to keep it _out_ of Redux. Using a store would make this easier to debug as well. And might improve postmortem debugging when/if we add Redux breadcrumbs to our Sentry reports.

Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:34]: It seems to be overkill, to me, to place something in redux that will only ever be used in a single view. Redux should be for the application-wide state, as per the redux-toolkit style guide recommendation [evaluate where each piece of state should live](https://redux.js.org/style-guide/style-guide#evaluate-where-each-piece-of-state-should-live)

Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:34]: Yeah, I guess. But, is this less complex than putting it in Redux? I saw that section as talking about when to use component state v.s. Redux, because in general keeping state in components is simpler if it concerns only those components. But creating intermediary reducers seems like _a lot_.

Maybe not though. I'll take another read through - I don't understand the context of this very well yet.

Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:34]: It actually started out as a use state, but the update logic was messy requiring a bunch of object spreads. Immer was a spur of the moment choice but it resulted in code that could very well be moved back to a useState hook instead of useReducer. Three things made me think to put in an intermediary useReducer instead of useState/redux

1. The state of this page will not be needed anywhere else. The state on this page is determined by the templates that are rendered, and would have to be loaded into redux on each new confirmation only for it to be immediately consumed at the page level.
2. I needed an object which React recommends breaking multiple keys into multiple useState but that doesn't work with the programmatic nature of this state.
3. It fits the paradigm of a reducer and the update logic was messy in a setState call before immer

File: ui/app/components/app/confirmation-footer/confirmation-footer.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  alerts,
+}) {
+  const className = classnames('confirmation-footer', {
+    'confirmation-footer--with-alerts':
+      alerts && Object.keys(alerts).length > 0,
+  });
+
+  return (
+    <div className={className}>
+      {alerts && alerts}
```
Comment (danjm) [inline ui/app/components/app/confirmation-footer/confirmation-footer.js:20]: is `alerts && alerts` needed? is `{alerts}` sufficient?

File: ui/app/components/app/confirmation-footer/confirmation-footer.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      <div className="confirmation-footer__actions">
+        <Button
+          rounded
+          type="secondary"
+          onClick={onCancel}
+          onKeyUp={(event) => {
+            if (event.key === 'Enter') {
```
Comment (danjm) [inline ui/app/components/app/confirmation-footer/confirmation-footer.js:27]: This is for accessibility, correct? Not for this PR, but in the future we could build this functionality into the button component, so that we don't need to add an `onKeyUp` property to every button

File: ui/app/components/app/network-display/network-display.js [hunk 5]
```diff
@@ -45,7 +52,9 @@ export default function NetworkDisplay({
             size={indicatorSize}
             type={ColorIndicator.TYPES.FILLED}
             iconClassName={
-              networkType === NETWORK_TYPE_RPC ? 'fa fa-question' : undefined
+              networkType === NETWORK_TYPE_RPC && indicatorSize !== SIZES.XS
```
Comment (danjm) [inline ui/app/components/app/network-display/network-display.js:55]: Why does the indicator size affect whether or not `fa-question` is used as the icon? Is there no question icon at the extra small size?

Comment (brad-decker) [reply ui/app/components/app/network-display/network-display.js:55]: It looks pretty bad at this size, scrunched and distorted

Comment (brad-decker) [reply ui/app/components/app/network-display/network-display.js:55]: Also the design does not have one

File: ui/app/pages/home/home.component.js [hunk 6]
```diff
@@ -106,6 +109,8 @@ export default class Home extends PureComponent {
       history.push(CONFIRM_TRANSACTION_ROUTE);
     } else if (Object.keys(suggestedTokens).length > 0) {
       history.push(CONFIRM_ADD_SUGGESTED_TOKEN_ROUTE);
+    } else if (pendingApprovals.length > 0) {
```
Comment (Gudahtt) [inline ui/app/pages/home/home.component.js:112]: We had talked about this going ahead of other confirmations in one of the design calls, but.... on second thought, I think putting it last makes the most sense, as you've done here. This gives the user the opportunity to clear their queue of confirmations first. But more importantly, a dapp can't batch further actions after the network switch anyway because we don't support multi-network connections yet. So this will always be the last action a dapp proposes.

File: ui/app/components/app/confirmation-footer/confirmation-footer.js [hunk 7]
```diff
@@ -0,0 +1,57 @@
+import React from 'react';
```
Comment (Gudahtt) [inline ui/app/components/app/confirmation-footer/confirmation-footer.js:1]: Would this make more sense underneath the `confirmation` page directory? I doubt this will be used elsewhere.

File: ui/app/pages/confirmation/confirmation.js [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+  /**
+   * Computation of the current alert state happens every time the current
+   * pendingConfirmation changes. The async function getTemplateAlerts is
+   * responsible for returning alert state. Setting state on unmounted
+   * components is an anti-pattern, so we use a signal variable to keep track
+   * of the current state of the component. Returning a function that sets
+   * signal to false when the component is unmounted.
+   */
+  useEffect(() => {
+    let signal = true;
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/confirmation.js:88]: Nit: Why "signal"? :thinking: I've seen something like this done before using an `isMounted` state. Maybe `isMounted` would be a better name for this?

File: ui/app/pages/confirmation/confirmation.js (hunk 2 above)
Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:34]: The reason I consider this as "application-wide state", because the alert state is interwoven with _a lot_ of other things, such as the confirmation properties, selected address, selected network, eth balance, token balance, etc. I see that this isn't handled in this reducer right now, but rather as part of the template `getAlerts` function. But this is exactly the sort of thing Redux is well suited for handling.

This reducer is simple and lacking any interaction with application-wide state at the moment only because the complex state logic is in `getAlerts`.

Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:34]: It was _intentional_ to isolate it from redux, though, because third-party developers in the future will presumably not have access to the redux store. If we did just provide a filtered state tree from redux to the `getAlerts` method provided by third parties, it seems safer to isolate the returned state from the rest of our application state.

I would contest that simply accessing application-wide state is not enough of a signal to warrant any derived state also presiding in the application-wide state.

I am not really following what you're suggesting here. Are you suggesting removing/significantly rework the `getAlerts` API for the template system?

File: app/_locales/en/messages.json [hunk 9]
```diff
@@ -148,7 +165,7 @@
     "message": "Approval Tx Gas Cost"
   },
   "approve": {
-    "message": "Approve spend limit"
+    "message": "Approve"
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/_locales/en/messages.json:168]: Yeah, fwiw the work on this predated to your changes to make the token category calls much clearer/more explicit. I had thought this was unused but see the err in my ways now.

File: ui/app/pages/confirmation/confirmation.js [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+        <Box justifyContent="center">
+          <NetworkDisplay
+            colored={false}
+            indicatorSize={SIZES.XS}
+            labelProps={{ color: COLORS.BLACK }}
+          />
+        </Box>
+        <Box justifyContent="center" padding={[1, 4, 4]}>
+          <Chip
+            label={stripHttpSchemes(originMetadata.origin)}
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:201]: This was an unconscious change in logic. I agree that only HTTPS should be stripped. I misread the method name + didn't check the function definition.

File: ui/app/pages/confirmation/confirmation.js (hunk 8 above)
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:88]: isMounted works, I was mimicking a useCancelablePromise hook I've written in the past when I wrote this. I'll reword it

File: ui/app/pages/confirmation/confirmation.scss [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    'content'
+    'footer';
+
+  &__content {
+    grid-area: content;
+    scrollbar-width: none;
+    padding: 16px 16px 0;
+
+    &::-webkit-scrollbar {
+      display: none;
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.scss:21]: Leftover code from when the footer buttons were sticky to the bottom of the page, will remove

File: ui/app/pages/confirmation/confirmation.js [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+            labelColor="gray"
+            leftIcon={
+              <SiteIcon
+                icon={originMetadata.icon}
+                iconName={originMetadata.name}
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:206]: Where in the designs are you seeing this? It's only used as a fallback when the icon isn't supplied. I'm not questioning you just can't find it to confirm the right behavior and resolve.

File: ui/app/pages/confirmation/confirmation.js (hunk 2 above)
Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:34]: Right, I hadn't considered third-party templates. I was thinking of this as another of our pages.

Where I was coming from with this is that I've been trying to ensure we keep our front-end components as simple as possible and move UI application logic into Redux where possible. I'd even say that too much logic in our components has historically been the a big part of _most_ of the technical debt in this project. So I'm having trouble envisioning how we'll keep that from happening again with these templates as the UI managed with templates gets increasingly complex.

File: ui/app/pages/confirmation/confirmation.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:206]: Sorry, when I said "according to the latest designs" I meant that the `name` was _absent_ from all of our _current_ Figma designs. I don't believe it has been documented anywhere that it should not be used. Though this has been touched on once or twice recently during design syncs I think.

What I recall about the site icon from the 'LoginPerSite" designs is that the fallback icon name should be the hostname. Even for HTTP, so that the fallback icon isn't "H" for everything. But again I'm unsure whether this is documented anywhere unfortunately.

The basically reason we had for not using `originMetadata.name` is that we noticed that in practice it was full of junk. We get it from the site title, and dapps tend to put all sorts of nonsense there, not just their names.

File: ui/app/pages/confirmation/confirmation.js (hunk 2 above)
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:34]: Well, it definitely isn't my intention to _create technical debt_. I think that we should probably define some of these concepts in a formal, written, and unanimously approved engineering principles/design document that mirrors the redux toolkit one. In our document I'd love for us to have a more clear definition of complexity as it relates to [move complex logic outside components](https://redux.js.org/style-guide/style-guide#move-complex-logic-outside-components). Having a hive mind about how we approach implementation as much as possible is going to be super important as we rapidly scale our team. 

For what it's worth I do see that it's impossible to not cross the line of complexity as more confirmations move to use the template system. I tackled this project with the concept of having this work for third-party consumers, without thinking about how we could best utilize the same overall system.

I do think that as we tackle the send confirmation or any of the others that the use cases for the state within the confirmation page are going to expand. I'd like to take time to think of a store structure that would allow many permutations of state for the confirmation page and implement it as part of the next confirmation that gets worked into this system.

Would that be agreeable to you, to address this concern as I work through the next confirmation implementation?

File: ui/app/pages/confirmation/confirmation.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js:206]: Thanks! so just to recap:
- [ ] name should just be hostname in useOriginMetadata (your suggestion that I can just commit now)
- [ ] the name prop on siteIcon should use hostname.
- [ ] change this behavior anywhere <SiteIcon> is consuming origin Metadata to use hostname instead?

File: ui/app/pages/home/home.component.js (hunk 6 above)
Comment (brad-decker) [reply ui/app/pages/home/home.component.js:112]: Woo hoo. mistakes that work in my favor for the win.

File: ui/app/pages/confirmation/confirmation.js (hunk 2 above)
Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:34]: Absolutely, that sounds like a good plan :+1:

File: ui/app/pages/confirmation/confirmation.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/confirmation/confirmation.js:206]: Sounds right to me, yep!

File: ui/app/pages/confirmation/confirmation.js (hunk 1 above)
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js]: Resolved in latest commit

File: ui/app/components/app/confirmation-footer/confirmation-footer.js (hunk 3 above)
Comment (brad-decker) [reply ui/app/components/app/confirmation-footer/confirmation-footer.js]: resolved in latest commit

File: app/_locales/en/messages.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply app/_locales/en/messages.json:171]: resolved in latest commit.

File: ui/app/hooks/useOriginMetadata.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+export function useOriginMetadata(origin) {
+  const domainMetaData = useSelector(getDomainMetadata);
+  if (!origin) {
+    return null;
+  }
+  const url = new URL(origin);
+
+  const minimumOriginMetadata = {
+    host: url.host,
+    name: url.hostname,
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/hooks/useOriginMetadata.js]: resolved in latest commit.

File: ui/app/pages/confirmation/confirmation.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js]: resolved in latest commit.

File: ui/app/components/app/confirmation-footer/confirmation-footer.js (hunk 7 above)
Comment (brad-decker) [reply ui/app/components/app/confirmation-footer/confirmation-footer.js]: resolved in latest

File: ui/app/pages/confirmation/confirmation.js (hunk 8 above)
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js]: resolved in latest commit.

File: ui/app/store/actions.js [hunk 14]
```diff
@@ -2463,6 +2463,40 @@ export function clearPermissions() {
   };
 }
 
+// Pending Approvals
+
+/**
+ * @param {string} id
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/store/actions.js]: resolved in latest

File: ui/app/pages/confirmation/confirmation.scss (hunk 11 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.scss]: removed in latest

File: ui/app/pages/confirmation/confirmation.scss [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+
+  &__navigation &__navigation-button:last-child {
+    margin-left: 8px;
+  }
+
+  &__origin-chip {
+    margin: 0 auto 16px auto;
+  }
+
+  & .confirmation-footer {
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.scss]: this became unnecessary without trying to treat the confirmation-footer as reusable. This style is now just directly in the confirmation-footer code.

File: ui/app/pages/confirmation/confirmation.scss [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+    grid-area: footer;
+  }
+
+  &__view-all-details {
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.scss]: removed

File: ui/app/pages/confirmation/confirmation.scss [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+      background-color: $Grey-100;
+      color: $Grey-300;
+    }
+  }
+
+  &__navigation &__navigation-button:last-child {
+    margin-left: 8px;
+  }
+
+  &__origin-chip {
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.scss]: removed

File: ui/app/pages/confirmation/templates/switch-ethereum-chain.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+    onCancel: () =>
+      actions.rejectPendingApproval(
+        pendingApproval.id,
+        ethErrors.provider.userRejectedRequest(),
+      ),
+  };
+}
+
+const addEthereumChain = {
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/templates/switch-ethereum-chain.js:91]: I'm guessing this was meant to be `switchEthereumChain`?

File: ui/app/pages/confirmation/templates/index.js [hunk 19]
```diff
@@ -0,0 +1,122 @@
...
+    rejectPendingApproval: (...args) =>
+      dispatch(rejectPendingApproval(...args)),
+    resolvePendingApproval: (...args) =>
+      dispatch(resolvePendingApproval(...args)),
+  };
+}
+
+/**
+ * Returns the templated values to be consumed in the confirmation page
+ * @param {Object} pendingApproval - The pending confirmation object
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/templates/index.js:96]: Nit: It would be nice to have types for `pendingApproval` as it's being passed around. I had to backtrack a few times to make sure I understood what it's properties were expected to be.

File: ui/app/pages/confirmation/templates/index.js [hunk 20]
```diff
@@ -0,0 +1,122 @@
...
+export async function getTemplateState(pendingApproval) {
+  const fn = APPROVAL_TEMPLATES[pendingApproval.type]?.getState ?? emptyState;
+  const result = await fn(pendingApproval);
+  if (typeof result !== 'object' || Array.isArray(result)) {
+    throw new Error(`Template state must be an object, received: ${result}`);
+    // I'm using == here to catch undefined and null because we'll want
+    // to enforce the object return shape in the event that one of these
+    // values is returned.
+    // eslint-disable-next-line no-eq-null
+  } else if (result == null) {
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/templates/index.js:71]: Nit: It might be simpler to check both results explicitly, rather than disable this lint rule and rely upon the weird semantics of `==`, then leave a long comment explaining it. Though I do appreciate the explanation!

```suggestion
  } else if (result === null || result === undefined) {
```

**Comment by Gudahtt** [comment]
I noticed that the network display and dapp origin chips were focusable for some reason :thinking: They had the role "button" despite not being clickable. However the "View all details" link wasn't focusable, nor were the tooltip info-icons.

Ideally the non-interactive components would be focusable, including the info-icon so that keyboard users can focus them to make the tooltip appear.

File: ui/app/pages/confirmation/confirmation.js [hunk 21]
```diff
@@ -0,0 +1,239 @@
...
+          <NetworkDisplay
+            colored={false}
+            indicatorSize={SIZES.XS}
+            labelProps={{ color: COLORS.BLACK }}
+          />
+        </Box>
+        <Box justifyContent="center" padding={[1, 4, 4]}>
+          <Chip
+            label={stripHttpsScheme(originMetadata.origin)}
+            labelColor="gray"
```
Comment (Gudahtt) [inline ui/app/pages/confirmation/confirmation.js:202]: I don't see a `labelColor` prop :thinking: Was this meant to be `borderColor`, or a `labelProp` perhaps?

File: ui/app/components/app/network-display/network-display.js [hunk 22]
```diff
@@ -78,6 +87,13 @@ NetworkDisplay.propTypes = {
   labelProps: PropTypes.shape({
     ...Chip.propTypes.labelProps,
```
Comment (Gudahtt) [inline ui/app/components/app/network-display/network-display.js:88]: I think this proptype is messed up :thinking: 

I keep seeing this error in the console: 
```
Warning: Failed prop type: The prop `labelProps.isRequired` is marked as required in `NetworkDisplay`, but its value is `undefined`.
    in NetworkDisplay (created by ConfirmationPage)
    in ConfirmationPage (created by Context.Consumer)
    in Route (created by Authenticated)
    ...
```

Presumably introduced in a different PR so maybe out of scope, but I'm just mentioning because it kept catching my eye in testing.

**Comment by Gudahtt** [comment]
The two links on the "Add chain" confirmation were the same colour as the surrounding text. This made them difficult to identify as being links. I only noticed they were clickable because they were focusable. In the designs they were a light blue colour.

**Comment by Gudahtt** [comment]
I noticed that I could call this method prior to connecting the dapp. This method should be callable only by connected dapps I would assume :thinking:

File: ui/app/pages/confirmation/templates/add-ethereum-chain.js [hunk 23]
```diff
@@ -0,0 +1,214 @@
...
+      matchedChain.name.toLowerCase() !==
+        pendingApproval.requestData.chainName.toLowerCase() ||
+      matchedChain.nativeCurrency?.symbol !== pendingApproval.requestData.ticker
+    ) {
+      validated = false;
+    }
+
+    const { hostname } = new URL(pendingApproval.requestData.rpcUrl);
+    if (
+      !matchedChain.rpc.map((rpc) => new URL(rpc).hostname).includes(hostname)
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js:91]: I could match the scheme too

File: ui/app/pages/confirmation/templates/add-ethereum-chain.js [hunk 24]
```diff
@@ -0,0 +1,214 @@
...
+  );
+  const matchedChain = safeChainsList.find(
+    (chain) =>
+      chain.chainId === parseInt(pendingApproval.requestData.chainId, 16),
+  );
+  let validated = Boolean(matchedChain);
+
+  if (matchedChain) {
+    if (
+      matchedChain.nativeCurrency?.decimals !== 18 ||
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js:81]: yeah, the only value we accept is 18 and so if the matched chain returns a different number than that we already know we have a mismatch.

File: ui/app/pages/confirmation/templates/add-ethereum-chain.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js:91]: I don't really have a preference - I was just digging for an interesting story to be honest :stuck_out_tongue:

[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js:91]: I'm afraid you're going to be disappointed cause this ended with me googling "parts to a URL" and picking the one that made the most sense to me when putting the check in place haha. We only allow https on our side, but chainid.network has HTTP:// providers listed. The main reason I broke apart the URL was to try and not match against things like API keys etc for Infura. probably worth while to match scheme + hostname.

[the remark this answers is not in the document]
Comment (Gudahtt) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js:91]: Gotcha - that makes sense.

If you're going to change it anyway I'd suggest going a step further and using the origin, to include the port as well. The origin is the basic unit of trust we use in most places, so it seems appropriate. Oddly enough, quite a few entries in the list include non-standard ports.

File: ui/app/components/app/network-display/network-display.js (hunk 22 above)
Comment (brad-decker) [reply ui/app/components/app/network-display/network-display.js:88]: I figured it out, lol. that particular prop (labelProps) is already a shape, it didn't need to be spread into a new shape. that caused....irregularities. i'm surprised it even worked without a more interesting error!

**Comment by brad-decker** [comment]
>I noticed that when confirming on the "Switch networks" confirmation page, I see this error in the dapp:

I saw this but was able to trigger it outside of the switch network confirmation. Open the test dapp, then open the extension in full-screen view in another tab. Watch the console as you try and switch to custom RPC endpoints. I switched between Huobi testnet/mainnet and xDAI, and occasionally I'd get a 'loading' spinner/view on the extension. Anytime I got this, I'd get the error in the test dapps console.

Furthermore, it seems intermittent in the same way when using the switch network confirmation, so I don't think this is something we need to solve in this PR. Perhaps it does warrant additional investigation, though.

> I noticed that I could call this method prior to connecting the dapp. This method should be callable only by connected dapps I would assume 🤔

I kind of feel like we should be allowing dapps to suggest chains before they interact with the user's accounts, for the case where dapps do not interact with a user through mainnet at all? I imagine if we do have a future where we have separate accounts per chain, forcing the dapp to make a connection to the user's mainnet/whatever chain to suggest another chain be added seems strange.

File: ui/app/pages/confirmation/templates/add-ethereum-chain.js (hunk 23 above)
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/templates/add-ethereum-chain.js]: changed to origin 👍

File: ui/app/pages/confirmation/confirmation.js [hunk 25]
```diff
@@ -0,0 +1,239 @@
...
+  if (!pendingConfirmation) {
+    return null;
+  }
+
+  return (
+    <div className="confirmation-page">
+      {pendingConfirmations.length > 1 && (
+        <div className="confirmation-page__navigation">
+          <p>
+            {currentPendingConfirmation + 1} of {pendingConfirmations.length}{' '}
```
[the remark this answers is not in the document]
Comment (brad-decker) [reply ui/app/pages/confirmation/confirmation.js]: i added `xOfYPending` to handle this.

**Comment by Gudahtt** [comment]
>I saw this but was able to trigger it outside of the switch network confirmation.

Huh, OK! I hadn't realized.

>I kind of feel like we should be allowing dapps to suggest chains before they interact with the user's accounts

Yeah maybe :thinking: In any case, we can adjust this in some later PR if we decide to restrict it.

**Review by Gudahtt — APPROVED** [review]
LGTM!

I had a couple of more comments (font size too small, and the origin shouldn't be greyed out/de-emphasized), but I left them in the Figma since they were done according to the design. We can figure that out in a subsequent PR.

I've tested this branch manually again, and everything seems to work great! It's especially great to see that keyboard navigation works perfectly - this may be the first confirmation that's true of.
